### DISCUSSION TITLE
Custom Applicative Metrics: Collect the number of orders created in a period of time

### DISCUSSION THREAD

**Discussion by ziomill** [open-telemetry/opentelemetry-java#4634] [Q&A]
Hello,

Suppose you have an e-commerce platform and you want to collect statistics about how many ORDERS are created IN A CERTAIN PERIOD OF TIME (ex. every day):

![ecommerce_stats](https://user-images.githubusercontent.com/22731494/180756196-c8286243-7ea9-49d4-b4ab-2bcadb6c0170.PNG)

Is it right to use a "Histogram" as explained [here ](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#histogram-creation)  ? 

When the **telemetry data** should be recorded? I have thought to implement a scheduled process which extracts the stats every day at midnight and uses the DoubleHistogram.record() in order to send the telemetry data. Is this the right approach?

Thank you

**Comment by jkwatson** [comment]
If you're only reporting a single datapoint a day, I'd say that OTel metrics might be overkill for what you're trying to measure. But, if you really want to use OTel metrics for this, how about something simple like an UpDownCounter?

**Reply by ziomill** [reply]
Hi @jkwatson 
I'm playing around with OpenTelemetry, and in particular with OpenTelemetry API, in order to better understand the concepts.
Anyway, I will deepen the UpDownCounter meter.
Thank you!

**Reply by jack-berg** [reply]
FYI, up down counter is only correct if you intend to to record negative values to the the instrument. Intuitively, I would think that you would increment the number of created orders, but never decrement it.

**Comment by jack-berg** [comment]
Histograms are used when the distribution of recorded measurements is of interest. For example, you'd use a histogram to record the response time of an http server fulfilling requests, since its interesting to not only see how many requests fulfilled, but also the distribution of time so you can compute response time percentiles. 

In your case, it _appears_ you're only interested in the how many orders were created. If this is the case, a `counter` instrument will suffice. There's an example [here](https://opentelemetry.io/docs/instrumentation/java/manual/#metrics) you can take a look at. In terms of _when_ the data should be recorded, you have a couple of options: you could increment the instrument synchronously when an order is created, you could have a batch job running each night recording the total number of orders created for the day, you could have an _asynchronous_ counter registered with a callback which is invoked periodically and records the total number of orders all time.

**Reply by ziomill** [reply]
Hi @jack-berg , 
The _async counter_ may seem the most elegant solution for this use case.
I would deepen this solution.
Thank you

**Reply by ziomill** [reply]
Hi @jack-berg ,
I'm trying the async counter and it seems to be exactly what I need.

I see that the _callback function_ passed to the _ObservableLongCounter_ is called every minute. Is there a way to customize this scheduling time?

**Reply by jack-berg** [reply]
The callback functions are invoked each time a `MetricReader` reads metrics. If you're using the `PrometheusHttpServer` metric reader, this will happen each time the metrics endpoint is requested. If you're using `PeriodicMetricReader` with a push based `MetricExporter`, the read interval can be set via `PeriodicMetricReaderBuilder#setInterval`.

**Reply by ziomill** [reply]
Ok, I found it ( from https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure):

![image](https://user-images.githubusercontent.com/22731494/181777345-04dc59dd-f258-4d0e-b337-54644d393465.png)

@jack-berg thank you for your suggestions!
